src/libpgagroal/utils.c

```diff
@@ -954,128 +955,124 @@ pgagroal_backtrace(void)
+         return false;
       }
-
-      // do I need also a value?
-      if (value)
+      else
       {
-         offset++;
-         *value = argc > offset ? argv[offset] : default_value;
-
-         if (*value == NULL || strlen(*value) == 0)
+         offset--;
+         subcommand = "";
+         for (int i = 0 ; i < command_match_count; i++)
          {
-            goto error;
+            if (strncmp(subcommand, command_table[command_match[i]].subcommand, MISC_LENGTH) == 0)
+            {
+               cmd->command_code = command_match[i]; /* pushes down the matches */
+            }
+         }
+         if (cmd->command_code == COMMAND_UNKNOWN)
+         {
+            sprintf(cmd->error_message, "Invalid subcommand '%s' for command '%s'\n", subcommand, command);
+            return false;
          }
-
       }
    }
 
-   return true;
-
-error:
-   return false;
-}
-
-bool
-parse_deprecated_command(int argc,
-                         char** argv,
-                         int offset,
-                         char* command,
-                         char** value,
-                         char* deprecated_by,
-                         unsigned int deprecated_since_major,
-                         unsigned int deprecated_since_minor)
-{
-   // sanity check: if no arguments, nothing to parse!
-   if (argc <= offset)
+   /* Iterate until finds an accepted_arg_count that is equal or greater than the typed command arg_count */
+   cmd->arg_count = argc - offset;
+   int j;
+   for (j = 0; j < MISC_LENGTH && command_table[cmd->command_code].accepted_argument_count[j] < cmd->arg_count; j++)
+   {
+      ;
+   }
+   if (cmd->arg_count < command_table[cmd->command_code].accepted_argument_count[0])
    {
+      sprintf(cmd->error_message, "Too few arguments provided for command '%s%s%s'\n", command,
+              (command && !EMPTY_STR(subcommand)) ? " " : "", subcommand);
       return false;
    }
-
-   // first of all check if the command is the same
-   // as the first argument on the command line
-   if (strncmp(argv[offset], command, MISC_LENGTH))
+   if (j == MISC_LENGTH || cmd->arg_count > command_table[cmd->command_code].accepted_argument_count[j])
    {
+      sprintf(cmd->error_message, "Too many arguments provided for command '%s%s%s'\n", command,
+              (command && !EMPTY_STR(subcommand)) ? " " : "", subcommand);
       return false;
    }
 
-   if (value)
+   /* Parse all arguments */
+   for (int i = 0; i < cmd->arg_count; i++)
    {
-      // need to evaluate the database or server
-      offset++;
-      *value = argc > offset ? argv[offset] : "*";
+      CONSUME(cmd->args[i]);
    }
 
-   // warn the user if there is enough information
-   // about deprecation
-   if (deprecated_by
-       && pgagroal_version_ge(deprecated_since_major, deprecated_since_minor, 0))
+   /* Warn the user if there is enough information about deprecation */
+   if (command_table[cmd->command_code].deprecated
```

Comment:
I'm wondering if we should remove some clutter of repeated `command_table[cmd->command_code]` by declaring a variable like `pgagroal_command template = command_table[ cmd->command_code];` and then use `template.deprecated`, `template.deprecated_by` and so on.

Another approach, could be to store a pointer to the entry in the command table into the command itself, so that it would be possible to `cmd->template->is_deprecated`. In this way, there would be no more need for a lookup into the table once the command has been successfully parsed.

Reply:
I like the second approach.